activerecord/lib/active_record/associations/join_dependency/join_association.rb

```diff
@@ -56,7 +56,7 @@ def join_constraints(foreign_table, foreign_klass, node, join_type, tables, scop
 
             klass_scope =
               if klass.current_scope
-                klass.current_scope.clone
+                remove_circular_references(klass.current_scope.clone)
```

Comment:
Oh, so I guess you mean joins_values are already used to build current join dependencies in this part and joins_values should not be used here. I think that's true. Fixed https://github.com/rails/rails/pull/25702/commits/487be17b8435ade20a4f6e04e73be584eccc9ade.